exec: null handling

jordanlewis opened the pull request:
@changangela, take a look at this - mainly just the change to colvec.go in the last commit.

jordanlewis committed:
util: add FastIntSet.CopyFrom

CopyFrom copies the contents of one FastIntSet to another, without
allocating new memory.

Release note: None

jordanlewis committed:
exec: null handling

ColVecs now track their nulls with a bitmap, and the cfetcher populates
the null bitmap properly.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32809)
<!-- Reviewable:end -->

RaduBerinde commented on pkg/util/fast_int_set.go:
```diff
@@ -200,6 +200,22 @@ func (s FastIntSet) Copy() FastIntSet {
 
+// CopyFrom sets the receiver to a copy of c, which can then be modified
+// independently.
+func (s *FastIntSet) CopyFrom(c FastIntSet) {
```
Why is this useful vs `a = b.Copy()`?

jordanlewis replied:
It's cheaper from what I recall. The extra allocations were showing in the profile heavily.

RaduBerinde replied:
Oh, I guess you were in the "large" case, where we are reusing the `Sparse` set.

RaduBerinde replied:
Note that `FastIntSet` is intended for cases where usually we are in the "small" case. For usecases where we are usually in the `large` case, using a `Sparse` directly or a slice of bits might be faster.

jordanlewis replied:
Ok, hmm. In this PR I added it to keep track of the "remaining values cols by idx" - in other words, the row-ordinal for every column that's expected to be in the value part of the keys within a row that we haven't seen so far. This needs to be tracked to figure out what bits to set in the per-column null bitmap.

I think these ordinals should be low in the common case - low being between 0 and 64, right?

RaduBerinde replied:
Yeah, I can't imagine that going over 64 in practice so perhaps the case where you were seeing allocs in `Copy` is worth investigating.

RaduBerinde replied:
(adding CopyFrom is fine with me though)

RaduBerinde commented:
Was the `util/fsm` library not useful for implementing the state machine?

jordanlewis commented:
@RaduBerinde, the state machine change is in a different PR - #32201. I took a brief look at using the util/fsm library but found its abstraction to be too heavyweight for my needs. I didn't know how to decompose the machine into events - the events in the fetcher are conditions that don't quite feel like they fit the event model. But perhaps I just wasn't looking at things in the right way... what do you think?

RaduBerinde commented:
Moved the conversation to that PR.

jordanlewis committed:
exec: add ColVec.PrettyValueAt

It returns a "pretty" string representing the value at an index in a
column.

Release note: None

changangela reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32809)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/colvec.go, line 137 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32809#-LU2_fg3-RvI0MzJQzIO:-LU2_fg3-RvI0MzJQzIP:b-ibo27a) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc45c315e6d3d7529f4760bdf41ddd3fb7ef8b1/pkg/sql/exec/colvec.go#L137)):*
> ```Go
> 
> func (m *memColumn) HasNulls() bool {
> 	sum := int64(0)
> ```

This might overflow and have weird behavior. Should we just use a bool instead?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32809)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/colvec.go, line 137 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32809#-LU2_fg3-RvI0MzJQzIO:-LU2a4jV8iySDfRvs7xv:b4k5zer) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc45c315e6d3d7529f4760bdf41ddd3fb7ef8b1/pkg/sql/exec/colvec.go#L137)):*
<details><summary><i>Previously, changangela (Angela Chang) wrote…</i></summary><blockquote>

This might overflow and have weird behavior. Should we just use a bool instead?
</blockquote></details>

Very true... :cold_sweat: 

I think we should use a bool as you suggest. But I'll leave that up to you - I'd like to get this merged so that you can at least be unblocked to work on it on your own :)


<!-- Sent from Reviewable.io -->

changangela reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32809)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/colvec.go, line 137 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32809#-LU2_fg3-RvI0MzJQzIO:-LU2aOu34LRjM0yUMjjA:brxp9sd) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc45c315e6d3d7529f4760bdf41ddd3fb7ef8b1/pkg/sql/exec/colvec.go#L137)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Very true... :cold_sweat: 

I think we should use a bool as you suggest. But I'll leave that up to you - I'd like to get this merged so that you can at least be unblocked to work on it on your own :)
</blockquote></details>

Sounds good!


<!-- Sent from Reviewable.io -->

jordanlewis commented:
bors r+
